controller/finalizer.go

```diff
@@ -79,27 +79,51 @@ func (f *Controller) addFinalizer(obj interface{}) (bool, error) {
 }
 
 func (f *Controller) removeFinalizer(obj interface{}) error {
-	patch, path, err := createRemoveFinalizerPatch(obj, f.name)
+	accessor, err := meta.Accessor(obj)
 	if err != nil {
 		return microerror.Mask(err)
 	}
-	if patch == nil {
+	if !containsFinalizer(accessor.GetFinalizers(), getFinalizerName(f.name)) {
+		// object has no finalizer set, this could have two reasons:
+		// 1. We are migrating and an old object never got reconciled before deletion.
+		// 2. The operator wasn't running and our first interaction with the object
+		// is its deletion.
+		// 3. The object has another finalizer set and we removed ours already.
+		// All cases should not be harmful in general, so we ignore it.
 		return nil
 	}
-	p, err := json.Marshal(patch)
-	if err != nil {
-		return microerror.Mask(err)
-	}
-	operation := func() error {
-		res := f.restClient.Patch(types.JSONPatchType).AbsPath(path).Body(p).Do()
```

Comment:
This was ugly beforehand, I didn't realize my mistake because there were two deletion events. The first one failed if another finalizer was removed at the same time, then the second one succeeded.

This should be fixed now.

Reply:
I do not follow here. Why are there two delete events? Which scenario is that? Also, why is the first event process concerned with the second one? This sounds like a race which we should not have at all within the same controller. When this is about having two controllers the race should already be managed by the resource version because one would get a conflict error. So this should already be fine, which makes me wonder what this PR is about. I miss something here.

Reply:
Yes you are right about the conflict and error. The problem I saw was that the conflict wasn't resolved and caused the `error rate` to go up.
This here fetches a fresh object with the current resourceVersion if an error like that conflict happens. 
So that instead of retrying with the old object and resourceVersion each time, we recover and don't wait for another delete event to remove the finalizer.

Reply:
Ok cool. That was not really clear to me based on this PR. Lets maybe add some comments to the function godoc what this is about and how it works and such.